pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+	}
+
+	if clusterCIDRSetList, ok := cidrMap[nodeSelector]; ok {
+		cidrMap[nodeSelector] = append(clusterCIDRSetList, clusterCIDRSet)
+	} else {
+		cidrMap[nodeSelector] = []*cidrset.ClusterCIDR{clusterCIDRSet}
+	}
+	return nil
+}
+
+// reconcileDelete deletes the ClusterCIDR object and removes the finalizer.
+func (r *multiCIDRRangeAllocator) reconcileDelete(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if slice.ContainsString(clusterCIDR.GetFinalizers(), clusterCIDRFinalizer, nil) {
+		if err := r.deleteClusterCIDR(clusterCIDR); err != nil {
+			return err
+		}
+		// Remove the finalizer as delete is successful.
+		cccCopy := clusterCIDR.DeepCopy()
+		cccCopy.ObjectMeta.Finalizers = slice.RemoveString(cccCopy.ObjectMeta.Finalizers, clusterCIDRFinalizer, nil)
+		if _, err := r.client.NetworkingV1alpha1().ClusterCIDRs().Update(context.TODO(), clusterCIDR, metav1.UpdateOptions{}); err != nil {
+			klog.V(2).Infof("Error removing finalizer for ClusterCIDR %s: %v", clusterCIDR.Name, err)
+			return err
+		}
+		klog.V(2).Infof("Removed finalizer for ClusterCIDR %s", clusterCIDR.Name)
+	}
+	return nil
+}
+
+// deleteClusterCIDR Deletes and unmaps the ClusterCIDRs from the cidrMap.
+func (r *multiCIDRRangeAllocator) deleteClusterCIDR(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+
+	labelSelector, err := r.nodeSelectorKey(clusterCIDR)
+	if err != nil {
+		return fmt.Errorf("unable to delete cidr: %w", err)
+	}
+
+	clusterCIDRSetList, ok := r.cidrMap[labelSelector]
+	if !ok {
+		klog.Infof("Label %s not found in CIDRMap, proceeding with delete", labelSelector)
+		return nil
+	}
+
+	for i, clusterCIDRSet := range clusterCIDRSetList {
+		if clusterCIDRSet.Name != clusterCIDR.Name {
+			continue
+		}
+
+		// Mark clusterCIDRSet as terminating.
+		clusterCIDRSet.Terminating = true
+
+		// Allow deletion only if no nodes are associated with the ClusterCIDR.
+		if len(clusterCIDRSet.AssociatedNodes) > 0 {
+			return fmt.Errorf("ClusterCIDRSet %s marked as terminating, won't be deleted until all associated nodes are deleted", clusterCIDR.Name)
+		}
+
+		// Remove the label from the map if this was the only clusterCIDR associated
+		// with it.
+		if len(clusterCIDRSetList) == 1 {
+			delete(r.cidrMap, labelSelector)
+			return nil
+		}
+
+		clusterCIDRSetList = append(clusterCIDRSetList[:i], clusterCIDRSetList[i+1:]...)
+		r.cidrMap[labelSelector] = clusterCIDRSetList
+		return nil
+	}
+	klog.V(2).Info("clusterCIDR not found, proceeding with delete", "Name", clusterCIDR.Name, "label", labelSelector)
+	return nil
+}
+
+func (r *multiCIDRRangeAllocator) nodeSelectorKey(clusterCIDR *networkingv1alpha1.ClusterCIDR) (string, error) {
+	var nodeSelector []byte
+	var err error
+
+	if clusterCIDR.Spec.NodeSelector != nil {
+		nodeSelector, err = clusterCIDR.Spec.NodeSelector.Marshal()
+	} else {
+		nodeSelector, err = defaultNodeSelector()
+	}
```

Comment:
I'm confused, the API says

```go
	// NodeSelector defines which nodes the config is applicable to.
	// An empty or nil NodeSelector selects all nodes.
	// This field is immutable.
	// +optional
	NodeSelector *api.NodeSelector
``` 

this returns 

```go
&v1.NodeSelector{
		NodeSelectorTerms: []v1.NodeSelectorTerm{
			{
				MatchExpressions: []v1.NodeSelectorRequirement{
					{
						Key:      defaultClusterCIDRKey,
						Operator: v1.NodeSelectorOpIn,
						Values:   []string{defaultClusterCIDRValue},
					},
				},
			},
		},
	}
```


is this `defaultClusterCIDRKey` an internal representation of the "All nodes"?
if that is the case you should add a more extensive explanation on the comment to make it more obvious

Reply:
Added a comment